File: Problems & Solutions/C/Case-specific Sorting of Strings/26 January 2023/Case-specific Sorting of Strings.java
//{ Driver Code Starts
//Initial Template for Java

import java.util.*;
import java.lang.*;
import java.io.*;


// } Driver Code Ends
//User function Template for Java

class Solution

{

    public static String caseSort(String str)

    {

        if(str.length()==1) return str;

        int[] cap = new int[26];

        int[] sma = new int[26];

        Arrays.fill(cap,0);

        Arrays.fill(cap,0);

        int z,y;

        for(int i=0;i<str.length();i++){

            z = (int)(str.charAt(i));

            if(z>=65 && z<=90){

                y = z-65;

                cap[y]++;

            }

            else{

                y = z-97;

                sma[y]++;

            }

        }

        ArrayList<Integer> capIndA = new ArrayList<>();

        ArrayList<Integer> smaIndA = new ArrayList<>();

        for(int i=0;i<26;i++){

            while(cap[i]-->0){

                capIndA.add(i);

            }

        }

        for(int i=0;i<26;i++){

            while(sma[i]-->0){

                smaIndA.add(i);

            }

        }

        StringBuffer t = new StringBuffer();

        char ch;

        int k=-1,j=-1;

        for(int i=0;i<str.length();i++){

            z = (int)(str.charAt(i));

            if(z>=65 && z<=90){

                k++;

                ch = (char)(capIndA.get(k)+65);

                t.append(ch);

            }

            else{

                j++;

                ch = (char)(smaIndA.get(j)+97);

                t.append(ch);

            }

        }

        return t.toString();

    }

}

//{ Driver Code Starts.

class GFG {
	public static void main (String[] args) {

    	try {
    	    BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
    	    PrintWriter out=new PrintWriter(System.out);
    	    String[] words = br.readLine().split("\\s+");
    	    int numTestCases = Integer.parseInt(words[0]);

    	    for (int tIdx = 0; tIdx < numTestCases; tIdx++) {
    	        words = br.readLine().split("\\s+");
    	        int size = Integer.parseInt(words[0]);
    	        String str = br.readLine();
    	        String sortedStr = new Solution().caseSort(str);
    	        out.println(sortedStr);
    	    }
    	    out.close();
    	}
    	catch (IOException e) {
    	    System.out.println(e);
    	}
	}
}
// } Driver Code Ends

File: Problems & Solutions/C/Case-specific Sorting of Strings/26 January 2023/Case-specific Sorting of Strings.js
//{ Driver Code Starts
//Initial Template for javascript

//Initial Template for javascript
'use strict';

process.stdin.resume();
process.stdin.setEncoding('utf-8');

let inputString = '';
let currentLine = 0;

process.stdin.on('data', inputStdin => {
    inputString += inputStdin;
});

process.stdin.on('end', _ => {
    inputString = inputString.trim().split('\n').map(string => {
        return string.trim();
    });
    
    main();    
});

function readLine() {
    return inputString[currentLine++];
}

function main() {
    let t = parseInt(readLine());
    let i = 0;
    for(;i<t;i++)
    {
        let n = parseInt(readLine());
        let string = readLine();
        
        let obj = new Solution();
        let ans = obj.caseSort(string, n);
        
        console.log(ans);
    }
}
// } Driver Code Ends

class Solution {
    caseSort(str, n) {
        let vecA = [], veca = [];
        
		for(let i=0;i<n;i++)
		{
			if(`${str.charCodeAt(i)}` < 97)
			vecA.push(str[i]);
			else
			veca.push(str[i]);
		}
		
		vecA.sort();
		veca.sort();
		
		let vcA=0,vca=0;
		let answer="";
		
		for(let i=0;i<n;i++)
		{
			if(`${str.charCodeAt(i)}` < 97)
			answer+=vecA[vcA++];
			
			else
			answer+=veca[vca++];
		}
		return answer;
    }
}

